gef.py

```diff
@@ -11301,112 +11090,503 @@ def screen_setup(self):
+                obj.cache_clear()
+            except: # we're reseeting the cache here, we don't care if (or which) exception triggers
+                continue
+        return
 
-    if sys.version_info[0] == 2:
-        err("GEF has dropped Python2 support for GDB when it reached EOL on 2020/01/01.")
-        err("If you require GEF for GDB+Python2, use https://github.com/hugsy/gef-legacy.")
+class GefMemoryManager(GefManager):
+    """Class that manages memory access for gef."""
+    def __init__(self) -> None:
+        self.reset_caches()
+        return
 
-    elif GDB_VERSION < GDB_MIN_VERSION:
-        err("You're using an old version of GDB. GEF will not work correctly. "
-            "Consider updating to GDB {} or higher.".format(".".join(map(str, GDB_MIN_VERSION))))
+    def reset_caches(self) -> None:
+        super().reset_caches()
+        self.__maps = None
+        return
 
-    else:
+    def write(self, address: int, buffer: ByteString, length: int = 0x10):  # -> memoryview object
+        """Write `buffer` at address `address`."""
+        return gdb.selected_inferior().write_memory(address, buffer, length)
+
+    def read(self, addr: int, length: int = 0x10) -> bytes:
+        """Return a `length` long byte array with the copy of the process memory at `addr`."""
+        return gdb.selected_inferior().read_memory(addr, length).tobytes()
+
+    def read_integer(self, addr: int) -> int:
+        """Return an integer read from memory."""
+        sz = gef.arch.ptrsize
+        mem = self.read(addr, sz)
+        unpack = u32 if sz == 4 else u64
+        return unpack(mem)
+
+    def read_cstring(self, address: int, max_length: int = GEF_MAX_STRING_LENGTH, encoding: Optional[str] = None) -> str:
+        """Return a C-string read from memory."""
+        encoding = encoding or "unicode-escape"
+        length = min(address | (DEFAULT_PAGE_SIZE-1), max_length+1)
+
+        try:
+            res_bytes = bytes(self.read(address, length))
+        except gdb.error:
+            err(f"Can't read memory at '{address}'")
+            return ""
+        try:
+            with warnings.catch_warnings():
+                # ignore DeprecationWarnings (see #735)
+                warnings.simplefilter("ignore")
+                res = res_bytes.decode(encoding, "strict")
+        except UnicodeDecodeError:
+            # latin-1 as fallback due to its single-byte to glyph mapping
+            res = res_bytes.decode("latin-1", "replace")
+
+        res = res.split("\x00", 1)[0]
+        ustr = res.replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")
+        if max_length and len(res) > max_length:
+            return f"{ustr[:max_length]}[...]"
+        return ustr
+
+    def read_ascii_string(self, address: int) -> Optional[str]:
+        """Read an ASCII string from memory"""
+        cstr = self.read_cstring(address)
+        if isinstance(cstr, str) and cstr and all([x in string.printable for x in cstr]):
```

Comment:
```suggestion
        if isinstance(cstr, str) and cstr and all(x in string.printable for x in cstr):
```

Reply:
No, this was already discussed: we keep the syntax `x in string.printable for x in cstr` (no square bracket)

Reply:
he's suggesting you remove the square brackets